packages/@stimulus/core/src/target_observer.ts

```diff
@@ -0,0 +1,50 @@
+import { Context } from "./context"
+import { Controller } from "./controller"
+import { TokenListObserver, TokenListObserverDelegate, Token, parseTokenString } from "@stimulus/mutation-observers"
+
+export class TargetObserver implements TokenListObserverDelegate {
+  readonly context: Context
+  readonly controller: Controller
+  readonly attributeName: string
+  private tokenListObserver: TokenListObserver
+
+  constructor(context: Context, controller: Controller) {
+    this.context = context
+    this.controller = controller
+    this.attributeName = `data-${context.identifier}-target`
+    this.tokenListObserver = new TokenListObserver(this.context.element, this.attributeName, this)
+  }
+
+  start() {
+    this.tokenListObserver.start()
+  }
+
+  stop() {
+    this.tokenListObserver.stop()
+  }
+
+  tokenMatched(token: Token): void {
+    if (this.controller.isConnected && this.containsDescendantWithToken(token.element, token.content)) {
+      this.dispatchCallback(`${token.content}TargetAdded`, token.element)
+    }
+  }
+
+  tokenUnmatched(token: Token): void {
+    if (this.controller.isConnected && !this.containsDescendantWithToken(token.element, token.content)) {
```

Comment:
Similarly, this `this.containsDescendantWithToken` is important here because we want to fire removal callbacks when the element itself is removed (i.e. `this.context.element.contains(token.element)` returns `false`), and when the element is _present_, but the token has been removed.